prompt2model/dataset_transformer/prompt_based.py

```diff
@@ -0,0 +1,123 @@
+"""A simple dataset transformer that uses a plan prompt and transform prompt."""
+
+from __future__ import annotations
+
+import asyncio
+
+import datasets
+
+from prompt2model.dataset_transformer.base import DatasetTransformer
+from prompt2model.dataset_transformer.prompt_template import (
+    construct_prompt_for_plan,
+    construct_prompt_for_transform_data,
+)
+from prompt2model.prompt_parser import PromptSpec
+from prompt2model.utils import (
+    API_ERRORS,
+    api_tools,
+    get_formatted_logger,
+    handle_api_error,
+)
+from prompt2model.utils.parse_json_responses import (
+    extract_response,
+    make_request_from_prompt,
+)
+
+logger = get_formatted_logger("DatasetTransformer")
+
+
+class PromptBasedDatasetTransformer(DatasetTransformer):
+    """A class for transforming given dataset to required format using a plan and a transform prompt."""  # noqa E501
+
+    def __init__(
+        self,
+        plan_prompt_fn=construct_prompt_for_plan,
+        transform_prompt_fn=construct_prompt_for_transform_data,
+    ):
+        """Initialize the class."""
+        self.plan_prompt_fn = plan_prompt_fn
+        self.transform_prompt_fn = transform_prompt_fn
+        self.plan = None
+
+    def canonicalize_dataset_using_samples(
+        self,
+        inputs: list[str],
+        outputs: list[str],
+    ) -> datasets.DatasetDict:
+        """Canonicalize a dataset into a suitable text-to-text format."""
+        assert len(inputs) > 0
+        assert len(inputs) == len(outputs)
```

Comment:
Asserts [should not be used for data validation](https://wiki.sei.cmu.edu/confluence/display/java/MET01-J.+Never+use+assertions+to+validate+method+arguments) in public functions.

```suggestion
        if len(inputs) == 0 or len(inputs) != len(outputs):
            raise ValueError(f"Length of inputs and outputs must be >0 and equal.")
```